docs/programming-guide.md

```diff
@@ -1086,6 +1086,29 @@ for details.
 </tr>
 </table>
 
+### Shuffle operations
+
+Certain operations within Spark trigger an operation known as the shuffle. The shuffle is Spark's mechanism for re-distributing data so that is grouped differently across partitions. This typically involves re-arranging and copying data across executors and machines, making shuffle a complex and costly operation. 
```

Comment:
I would replace uses of "operation" that don't refer to actions or transformations to avoid confusion.